fe/fe-core/src/main/java/com/starrocks/transaction/TransactionStmtExecutor.java

```diff
@@ -0,0 +1,499 @@
+            }
+
+            if (transactionState.getTableIdList().contains(targetTable.getId())) {
+                throw ErrorReportException.report(ErrorCode.ERR_TXN_IMPORT_SAME_TABLE);
+            }
+            transactionState.addTableIdList(targetTable.getId());
+
+            ExplicitTxnState.ExplicitTxnStateItem item =
+                    load(database, targetTable, execPlan, dmlStmt, originStmt, context, coordinator);
+            explicitTxnState.addTransactionItem(item);
+
+            context.getState().setOk(item.getLoadedRows(), Ints.saturatedCast(item.getFilteredRows()),
+                    buildMessage(transactionState.getLabel(), TransactionStatus.PREPARE,
+                            transactionState.getTransactionId(), database.getId()));
+        } catch (StarRocksException | RpcException | InterruptedException e) {
+            context.getState().setError(e.getMessage());
+        }
+    }
+
+    public static void commitStmt(ConnectContext context, CommitStmt stmt) {
+        ExplicitTxnState explicitTxnState = context.getExplicitTxnState();
+        if (explicitTxnState == null) {
+            //commit statement not in after begin, do nothing
+            return;
+        }
+
+        if (explicitTxnState.getTransactionStateItems().isEmpty()) {
+            TransactionState transactionState = explicitTxnState.getTransactionState();
+            context.setExplicitTxnState(null);
+            context.getState().setOk(0, 0, buildMessage(transactionState.getLabel(),
+                    TransactionStatus.VISIBLE, transactionState.getTransactionId(), -1));
+            return;
+        }
+
+        GlobalTransactionMgr transactionMgr = GlobalStateMgr.getCurrentState().getGlobalTransactionMgr();
+
+        long transactionId = context.getExplicitTxnState().getTransactionState().getTransactionId();
+        TransactionState transactionState = context.getExplicitTxnState().getTransactionState();
+        long databaseId = transactionState.getDbId();
+        Database database = GlobalStateMgr.getCurrentState().getLocalMetastore().getDb(databaseId);
+
+        try {
+            int timeout = context.getSessionVariable().getQueryTimeoutS();
+            long jobDeadLineMs = System.currentTimeMillis() + timeout * 1000L;
+
+            List<TabletCommitInfo> commitInfos = Lists.newArrayList();
+            List<TabletFailInfo> failInfos = Lists.newArrayList();
+            long loadedRows = 0;
+            for (ExplicitTxnState.ExplicitTxnStateItem item : explicitTxnState.getTransactionStateItems()) {
+                commitInfos.addAll(item.getTabletCommitInfos());
+                failInfos.addAll(item.getTabletFailInfos());
+                loadedRows += item.getLoadedRows();
+            }
+
+            TxnCommitAttachment txnCommitAttachment = new InsertTxnCommitAttachment(loadedRows);
+            VisibleStateWaiter visibleWaiter = transactionMgr.retryCommitOnRateLimitExceeded(
```

Comment:
can we ensure the "read-committed" isolation level this way?

Reply:
yes